### PR TITLE
Vulkan AS rebuild-on-replay: Copy AS input buffers on build

### PR DESCRIPTION
**Pull request by cmannett85-arm**
When `vkCmdBuildAccelerationStructuresKHR` is called, the input buffers used for the build are copied into host-accessible memory, so when initial state serialisation is triggered it can be copied into the capture file.

To unify the behaviour of the different geometry types, there is only one readback buffer per AS, and in the case of triangle data it can contain the vertex, index, and transform data concatenated into a single contiguous block.

The buffer data along with build arguments are stored in a `VkAccelerationStructureInfo` object owned by the AS's `VkResourceRecord`.  If this record is freed, then the copied buffer mem is also freed.  If this destruction occurs during the active capture phase, the AS info is kept alive until the end of the active capture.

Serialisation of the initial state will follow in later patches.

Caveats:
* `vkCmdBuildAccelerationStructuresIndirectKHR` is not handled
* `VkAccelerationStructureGeometryInstancesDataKHR::arrayOfPointers` mode is not handled

### CODE REVIEW

File: renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp [hunk 1]
```diff
@@ -335,6 +334,36 @@ void WrappedVulkan::vkDestroyImage(VkDevice device, VkImage obj, const VkAllocat
...
 }
 
+void WrappedVulkan::vkDestroyAccelerationStructureKHR(VkDevice device, VkAccelerationStructureKHR as,
+                                                      const VkAllocationCallbacks *)
+{
+  if(as == VK_NULL_HANDLE)
+    return;
+
+  // artificially extend the lifespan of the AS info to ensure it isn't destroyed before
+  // serialisation completes
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp]: I don't think this should be necessary because the capture process already refcounts the resource record to keep it alive for the duration of capture. Otherwise we would lose the chunks contained within. Anything that's marked resource referenced also adds a reference on the resource record. This doesn't protect against applications deleting resources referenced by a command buffer, but that would be illegal in vulkan (if the command buffer is later submitted).

Did you see this happening in practice or was this a theoretical fix?

File: renderdoc/driver/vulkan/vk_acceleration_structure.h [hunk 2]
```diff
@@ -28,9 +28,52 @@
...
+
+    struct Aabbs
+    {
+      VkDeviceSize stride;
+    };
+
+    struct Instances
+    {
+      // arrayOfPointers TODO
+    };
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.h]: I don't think we should have the BLASs referenced by a TLAS in the same way as geometry. This would be better stored as a `rdcarray<VkAccelerationStructureKHR> asData;` next to `geometryData` to more clearly indicate that only one is present.

File: renderdoc/driver/vulkan/vk_acceleration_structure.h [hunk 3]
```diff
@@ -28,9 +28,52 @@
...
+  ~VkAccelerationStructureInfo();
+
+  VkDevice device = VK_NULL_HANDLE;
+
+  VkAccelerationStructureTypeKHR type =
+      VkAccelerationStructureTypeKHR::VK_ACCELERATION_STRUCTURE_TYPE_GENERIC_KHR;
+  VkBuildAccelerationStructureFlagsKHR flags = 0;
+
+  rdcarray<GeometryData> geometryData;
+  rdcarray<VkAccelerationStructureBuildRangeInfoKHR> buildRangeInfo;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.h]: Rather than having two parallel arrays here I would prefer to have this information stored directly in `GeometryData` with one array.

I'm guessing maybe this is done so this separate build range array can be re-used as-is somewhere on build? but we still need to recreate the `VkAccelerationStructureBuildGeometryInfoKHR` array and the cost of reconstructing the build range array as a temporary will be mostly trivial in comparison - I think the ease of reading the code and seeing the related information together is worthwhile.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 4]
```diff
@@ -360,6 +757,136 @@ void VulkanAccelerationStructureManager::Apply(ResourceId id, const VkInitialCon
...
+  result.address = address - result.offset;
+  return result;
+}
+
+template <typename T>
+void VulkanAccelerationStructureManager::DeletePreviousInfo(VkCommandBuffer commandBuffer, T *info,
+                                                            CaptureState state)
+{
+  // If we're capturing this frame, then the info could be attached to an initial state so we need
+  // to delete it later
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: At the risk of sounding like a broken record, I think refcounting on the acceleration structure info would work better here for this kind of versioning/lifetime would work better here. I think it will also come in handy further down the line when we will likely want to start shuffling this data off to disk early and so the lifetime won't be as binary.

This also has the benefit of removing the need for the deferred cleanup process entirely I think.

You would need to lock the cap transition lock here, otherwise there's a potential significant race here, but generally I try to avoid checking the active/background capturing state where possible as it can lead to hard-to-test edge cases. At least when there's another fairly simple alternative.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 5]
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
...
+        {
+          VkMemoryRequirements mrq = {};
+
+          // Vertex buffer.  The complexity here is that the rangeInfo members are interpreted
+          // differently depending on whether or not index buffers are used
+          ObjDisp(device)->GetBufferMemoryRequirements(Unwrap(device), vertexData.buf, &mrq);
+          vertexData.alignment = mrq.alignment;
+          vertexData.size = vertexData.buf != VK_NULL_HANDLE
+                                ? vertexData.rao.record->memSize - vertexData.rao.offset
+                                : rangeInfo.primitiveCount * 3 * triInfo.vertexStride;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: I think this check is incorrect - `vertexData.buf` should never be NULL here - I think this meant to check if index data is present.

We have the information from `VkAccelerationStructureGeometryTrianglesDataKHR::maxVertex` for bounding the vertex buffer size in the case where index buffers are used, so we should use it otherwise we could copy a huge amount of extra data. We could `RDCMIN` these if we're concerned about people putting in a stupid value (which now that I think about it, someone definitely will).

Also for this code I think it's better to wrap both vertex + index parts in `if(indexData)` to make the code cleaner, even if it means a little duplication.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 6]
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
...
+            ObjDisp(device)->GetBufferMemoryRequirements(Unwrap(device), transformData.buf, &mrq);
+            transformData.alignment = mrq.alignment;
+            transformData.size = sizeof(VkTransformMatrixKHR);
+            transformData.start = rangeInfo.transformOffset;
+          }
+        }
+        const VkDeviceSize maxAlignment =
+            RDCMAX(RDCMAX(vertexData.alignment, indexData.alignment), transformData.alignment);
+
+        // We want to copy the ubput buffers into one big block so sum the sizes up together
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: typo: `input buffers`.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 7]
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
...
+        if(readbackmem.mem == VK_NULL_HANDLE)
+        {
+          RDCERR("Unable to allocate AS triangle input buffer readback memory (size: %u bytes)",
+                 totalMemSize);
+          continue;
+        }
+
+        // Insert copy commands
+        VkBufferCopy region = {
+            vertexData.rao.offset + vertexData.start,
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: I think there's a risk of confusion here in future with the terms `offset` and `start` being somewhat interchangeable. Here `offset` meaning "offset from the beginning of the actual buffer to the address passed in" and `start` being "the relative location where the geometry data sits".

Since you are using the local `BufferData` you could maybe rename these variables and add a helper function for this, but I wonder if you need to keep `start` separate? It looks like everywhere you use this it's always added to offset since once we've processed all the offsets and sizes here we really only care about the absolute offset from buffer start to where the data sits, the address the application passed in is mostly irrelevant.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 8]
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
...
+        // vertex buffer depending if indices are in use or not
+        VkAccelerationStructureBuildRangeInfoKHR buildData;
+        buildData.primitiveCount = rangeInfo.primitiveCount;
+        buildData.primitiveOffset =
+            !indexData ? 0 : (uint32_t)AlignUp(vertexData.size, vertexData.alignment);
+        buildData.firstVertex = !indexData ? 0 : rangeInfo.firstVertex;
+        buildData.transformOffset = !transformData
+                                        ? 0
+                                        : (uint32_t)(AlignUp(vertexData.size, vertexData.alignment) +
+                                                     AlignUp(indexData.size, indexData.alignment));
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: I think using `ifs` here would be easier to read than ternaries.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 9]
```diff
@@ -37,13 +37,410 @@ constexpr VkDeviceSize handleCountSize = 8;
...
+    }
+
+    // Insert barriers to block any other commands until the buffers are copied
+    if(!barriers.empty())
+    {
+      ObjDisp(device)->CmdPipelineBarrier(Unwrap(commandBuffer), VK_PIPELINE_STAGE_TRANSFER_BIT,
+                                          VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, 0, VK_NULL_HANDLE,
+                                          static_cast<uint32_t>(barriers.size()), barriers.data(),
+                                          0, VK_NULL_HANDLE);
+    }
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: I think rather than doing individual buffer barriers it would be easier to just do one global memory barrier. I'm not sure the exact difference for drivers but I believe few even use buffer barriers as anything specific, and if any do as soon as we have more than a couple of buffers here it's likely going to decay.

Also note that this barrier after the copies is to ensure that the input buffers aren't overwritten by any _later_ commands, so we should have both source and destination access masks to ensure that.

**Review by baldurk — CHANGES_REQUESTED** [review]
As a general note since I can't put this on the specific line, remember to disable the indirect build physical device feature.

It's fine not to support indirect builds as it seems few drivers do as well and I don't think anyone will use it given the feature is not supported in D3D12, but I think it should be technically possible as it looks like we get bounds on the data size on the CPU? The main challenge will be doing the copy, might have to fall back to a compute shader or something and do it inefficiently.

File: renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp (hunk 1 above)
Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp]: I did see it happening but given your explanation, perhaps I was misinterpreting.  I'll remove this and give your AS info refcounting idea a go, and that should solve the fiddly lifetime issues I was tackling.

File: renderdoc/driver/vulkan/vk_acceleration_structure.h (hunk 2 above)
Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: I don't follow what you're saying, this struct is meant to represent the non-memory state of `VkAccelerationStructureGeometryInstancesDataKHR` but I'm not handling `arrayOfPointers` hence why it is empty.  The actual `VkAccelerationStructureInstanceKHR` data is copied into `readbackMem` like everything else.

Comment (baldurk) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: Ah sorry I misunderstood what this is for.

With `arrayOfPointers` what we want to store isn't different we just can't do it in one copy. We'll probably need to do an inefficient compute-based copy to follow the indirections, but in that case this struct still wouldn't be needed as the data would still end up in `readbackMem` just via a more circuitous route?

Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: Ah yes, you're absolutely right as the copy would 'flatten' the indirection so on replay `arrayOfPointers` would always be false - I'll remove it.

Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: I'm going to need some hand holding when it comes to using a compute shader for the indirect copying - is there anything similar in the codebase already I can base it on?

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp (hunk 9 above)
Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: Is there any existing functionality that implements this?  Rather than me re-invent the wheel.

Comment (baldurk) [reply renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: It could be the same code as you have here but just one barrier at the end instead of N.

File: renderdoc/driver/vulkan/vk_acceleration_structure.h (hunk 2 above)
Comment (baldurk) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: I don't think there's anything similar on vulkan unfortunately. This is something uniquely difficult on RT needing this kind of GPU work at capture time.

Roughly it would mean tracking enough compute state in the command buffer during capture to push + pop around our own dispatch, and then write a shader to indirect through the pointers and copy out a BLAS-instance sized bit of memory to the output. I think since you will inevitably disturb the pipeline layout and push constants etc you can push-constant the number of BLASs.

Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: For this patch would it be acceptable to just fail the capture if `arrayOfPointers` is set and then return to the issue after the general case is merged?  I'm worried about getting bogged down in this when I haven't even seen the mode used yet.

Comment (baldurk) [reply renderdoc/driver/vulkan/vk_acceleration_structure.h]: Oh yes sorry I didn't mean to imply you needed to add support here. As long as there's an assert that the case isn't handled it can be left as a TODO for a future PR. I only made the original comment because I misunderstood what you were needing this struct for.

File: renderdoc/driver/vulkan/vk_acceleration_structure.h [hunk 10]
```diff
@@ -43,8 +87,35 @@ class VulkanAccelerationStructureManager
...
+  };
+
   explicit VulkanAccelerationStructureManager(WrappedVulkan *driver);
 
+  // Allocates readback mem and injects commands into the command buffer so that the input buffers
+  // are copied.
+  RDResult CopyInputBuffers(VkCommandBuffer commandBuffer,
+                            const VkAccelerationStructureBuildGeometryInfoKHR &info,
+                            const VkAccelerationStructureBuildRangeInfoKHR *buildRange,
+                            CaptureState state);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.h]: This state seems redundant here - we don't do this during replay, and we don't need to differentiate between background and active capturing. Doing a quick look through unless I'm missing something this value is (ultimately) not used.

Same applies below to `CopyAccelerationStructure`.

File: renderdoc/driver/vulkan/vk_resources.cpp [hunk 11]
```diff
@@ -3948,6 +3948,107 @@ void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
...
+  m_Entries.reserve(m_Entries.size() + values.size());
+  for(uint64_t value : values)
+    m_Entries.emplace_back(firstQuery++, value);
+
+  std::sort(m_Entries.begin(), m_Entries.end());
+}
+
+void QueryPoolInfo::Reset(uint32_t firstQuery, uint32_t queryCount)
+{
+  m_Entries.removeIf([&](const auto &entry) {
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_resources.cpp]: Please avoid `auto` except where necessary for lambdas or extremely ugly for std iterators, you can use `Entry` here.

Also please use explicit lambda capture rather than unconditional capture to be clear about what is stored in the lambda.

File: renderdoc/driver/vulkan/vk_resources.cpp [hunk 12]
```diff
@@ -3948,6 +3948,107 @@ void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
...
+  if(m_Entries.empty())
+    return {1, 0};    // Invalid
+
+  const uint32_t start = RDCMAX(firstQuery, m_Entries.front().index);
+  const uint32_t end = RDCMIN(firstQuery + queryCount - 1, (uint32_t)m_Entries.back().index);
+  return {start, end};
+}
+
+void QueryPoolInfo::Replace(uint32_t firstQuery, uint32_t queryCount,
+                            const std::function<void(Entry)> &writeEntry) const
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_resources.cpp]: Is there a reason for this function to be separated out? It looks like it's only called from the other `Replace()` to apply any replacements so could be done inside that function? That should allow some simplification and ease of reading since it took me a couple of passes to understand the code flow with the functions being named the same.

File: renderdoc/driver/vulkan/vk_resources.h [hunk 13]
```diff
@@ -2236,6 +2236,41 @@ inline FrameRefType MarkMemoryReferenced(std::unordered_map<ResourceId, MemRefs>
...
+    uint32_t index;
+    uint64_t value;
+  };
+
+  rdcpair<uint32_t, uint32_t> GetIntersection(uint32_t firstQuery, uint32_t queryCount) const;
+
+  void Replace(uint32_t firstQuery, uint32_t queryCount,
+               const std::function<void(Entry)> &writeEntry) const;
+
+  rdcarray<Entry> m_Entries;
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_resources.h]: Thinking about this more generally, query pools are single-typed so we know at creation time whether we're either going to replace all results or none of them (when the query pool type is compacted size).

Do you have any empirical data or intuition on how large those query pools are? I would assume either 1, some small fixed count (UE5 seems to use 64, configurable), or maybe the application's expected max number of BLASs (some thousands?).

Unless that seems unreasonable or you're otherwise worried about having a huge number of possible queries with only a few actually being valid, I think this code could be simplified a good deal by a) only allocating this struct when needed for patching and b) pre-allocating all the entries up front. Then you don't have to worry about intersecting, resizing this array on the fly, etc. We just read all our values out of it.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 14]
```diff
@@ -8009,27 +8049,35 @@ void WrappedVulkan::vkCmdWriteAccelerationStructuresPropertiesKHR(
...
+  if(queryType == VK_QUERY_TYPE_ACCELERATION_STRUCTURE_COMPACTED_SIZE_KHR)
+  {
+    auto &qpInfo = GetRecord(queryPool)->queryPoolInfo;
+
+    rdcarray<uint64_t> sizes;
+    sizes.reserve(accelerationStructureCount);
+    for(uint32_t i = 0; i < accelerationStructureCount; ++i)
+      sizes.push_back(GetRecord(pAccelerationStructures[i])->memSize);
+
+    qpInfo->Add(firstQuery, std::move(sizes));
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Unfortunately I've realised seeing this that there is an ordering/timeline constraint that I hadn't considered with this data. Because the properties write to a query and then the queries are resolved separately, these commands can be recorded and submitted out of order so we can't update a single authoritative fake query pool on the CPU timeline and update from there.

There may be a clever solution to this to resolve it on the CPU by deferring query updates and keeping multiple versions around, but I think I'd say we should solve it on the GPU to keep things simpler:

1. Create an internal buffer backing the query pool (ie. what the driver probably does on most GPUs) of equal size to its full set of results. It should probably be readback memory to be CPU-accessible because of 4 below. In theory in future we could have two buffers kept up to date in parallel, one for CPU reads and one for GPU reads, but I doubt that will be a problem.
2. This `vkCmdWriteAccelerationStructuresPropertiesKHR` writes the data via `vkUpdateBuffer` into that backing buffer - again effectively in parallel to what the driver is doing.
3. In `vkCmdCopyQueryPoolResults` we do a plain `vkCmdCopyBuffer` from backing buffer to user's results buffer.
4. In `vkGetQueryPoolResults` we CPU-read from the backing buffer.

It's a little bit uglier/more work than tracking purely on the CPU side but it keeps the timeline neatly in sync without having to mirror it with submission tracking and CPU-side versioning of data.

File: renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp [hunk 15]
```diff
@@ -1699,6 +1705,8 @@ void WrappedVulkan::vkResetQueryPool(VkDevice device, VkQueryPool queryPool, uin
   SERIALISE_TIME_CALL(
       ObjDisp(device)->ResetQueryPool(Unwrap(device), Unwrap(queryPool), firstQuery, queryCount));
 
+  GetRecord(queryPool)->queryPoolInfo->Reset(firstQuery, queryCount);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_misc_funcs.cpp]: This also misses handling of `CmdResetQueryPool` but I think with the other comments handling reset may not be needed at all.

File: renderdoc/driver/vulkan/vk_acceleration_structure.cpp [hunk 16]
```diff
@@ -37,13 +38,418 @@ constexpr VkDeviceSize handleCountSize = 8;
...
+        break;
+      }
+      case VK_GEOMETRY_TYPE_INSTANCES_KHR:
+      {
+        const VkAccelerationStructureGeometryInstancesDataKHR &instanceInfo =
+            geometry.geometry.instances;
+
+        if(instanceInfo.arrayOfPointers)
+          return RDResult(ResultCode::InternalError,
+                          "AS instance build arrayOfPointers unsupported");
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_acceleration_structure.cpp]: FYI there is a helper `RETURN_ERROR_RESULT` which can do this and also print the `RDCERR` for you, without the calling code needing to do that.

**Review by baldurk — CHANGES_REQUESTED** [review]
The input buffer copying looks good. I've made some notes about the query pool handling though - there's one wrinkle I hadn't thought of which makes it more complex, but hopefully some ways it can be simplified so it's not too bad overall.

File: renderdoc/driver/vulkan/vk_resources.cpp (hunk 12 above)
Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_resources.cpp]: This is called from both the public `Replace(..)` methods, to de-duplicate the logic.

File: renderdoc/driver/vulkan/vk_resources.h (hunk 13 above)
Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/vk_resources.h]: That's a good point, the largest I've seen is 64 but that was for UE5 demos.  I can't find anything in the spec regarding an upper bound so I've assumed the worst.  I've asked on the Vulkan Discord if anyone knows if the spec does actually define it.

If we make the reasonable assumption that it will be a few thousand max, then allocating the buffer (as described below) using whatever size is passed into `VkQueryPoolCreateInfo::queryCount` will definitely simplify the code.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 17]
```diff
@@ -7873,6 +7920,15 @@ void WrappedVulkan::vkCmdBuildAccelerationStructuresKHR(
 
       // Add to the command buffer metadata, so we can know when it has been submitted
       record->cmdInfo->accelerationStructures.push_back(GetRecord(geomInfo.dstAccelerationStructure));
+
+      const RDResult copyResult = GetAccelerationStructureManager()->CopyInputBuffers(
+          commandBuffer, geomInfo, ppBuildRangeInfos[i]);
+      if(copyResult != ResultCode::Succeeded)
+      {
+        m_LastCaptureError = copyResult;
+        RDCERR("%s", copyResult.message.c_str());
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: This error print is no longer needed with the `RETURN_ERROR_RESULT` macro (which prints it at the source, so better file/line in the log.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 18]
```diff
@@ -4546,6 +4546,53 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
...
+        };
+        vkCmdCopyBuffer(commandBuffer, qpInfo->m_Buffer.buf, destBuffer, 1, &region);
+      }
+      else
+      {
+        // Copy each result into the destination, converting if required
+        rdcarray<VkBufferCopy> regions;
+        regions.reserve(queryCount);
+        for(size_t i = 0; i < queryCount; ++i)
+          regions.push_back(VkBufferCopy{(firstQuery + i) * sizeof(uint64_t),
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Please don't use this altered form of construction. You should be able to drop it here and just use braces for a POD struct, or else you could make a temporary struct outside the loop and update members inside (or resize + set, if you prefer).

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 19]
```diff
@@ -4546,6 +4546,53 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
...
+                                         destOffset + (i * destStride), resultSize});
+
+        vkCmdCopyBuffer(commandBuffer, qpInfo->m_Buffer.buf, destBuffer, (uint32_t)regions.size(),
+                        regions.data());
+      }
+
+      if(flags & VK_QUERY_RESULT_WITH_AVAILABILITY_BIT)
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Unfortunately the availability bit indicates a _per-query_ availability - so `uint64 size; uint64 avail;` etc. If this flag is passed we'll need to go through the `else` path above to split up copies and interleave with updates.

Again though I would not expect anyone to actually use this as it's mostly for things like occlusion queries that aren't waited on or such, not a size that you really really need.

**Review by baldurk — CHANGES_REQUESTED** [review]

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp (hunk 19 above)
Comment (cmannett85-arm) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: > Again though I would not expect anyone to actually use this

Yeah but the spec says they can and if someone does they'll read past the end of the results array (because it's half the size they expect) and complain.

Comment (baldurk) [reply renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: Definitely, I just mean it's not a problem to put it in the suboptimal bucket with non-64bit results by doing separate copies/updates.

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 20]
```diff
@@ -4539,13 +4539,63 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
...
+    {
+      VkMemoryBarrier barrier = {
+          VK_STRUCTURE_TYPE_MEMORY_BARRIER,
+          NULL,
+          VK_ACCESS_TRANSFER_WRITE_BIT,
+          VK_ACCESS_TRANSFER_READ_BIT,
+      };
+      vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_TRANSFER_BIT,
+                           VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT, 0, 1, &barrier, 0, VK_NULL_HANDLE, 0,
+                           VK_NULL_HANDLE);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: One other minor thing here - we should call directly into the unwrapped versions of all these commands as necessary. Otherwise these will 'look like' application calls and will be serialised into the command buffers and show up in captures.

File: renderdoc/driver/vulkan/vk_resources.cpp [hunk 21]
```diff
@@ -3948,6 +3948,19 @@ void ImgRefs::Split(bool splitAspects, bool splitLevels, bool splitLayers)
   areLayersSplit = newSplitLayerCount > 1;
 }
 
+QueryPoolInfo::QueryPoolInfo(WrappedVulkan *driver, VkDevice device,
+                             const VkQueryPoolCreateInfo *pCreateInfo)
+{
+  m_Buffer.Create(driver, device, pCreateInfo->queryCount * 8, 1, GPUBuffer::eGPUBufferReadback);
```
Comment (baldurk) [inline renderdoc/driver/vulkan/vk_resources.cpp:3954]: Something I forgot before - this needs a call to `ResourceManager::SetInternalResource` to ensure this buffer is not considered a user buffer, even though it's created and wrapped. Otherwise if someone enables 'ref all resources' this buffer will show up in the capture visibly. See how these buffers are created in the text renderer for an example of how this is handled for our internal buffers during capture.

**Review by baldurk — CHANGES_REQUESTED** [review]

File: renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp [hunk 22]
```diff
@@ -4539,13 +4539,66 @@ void WrappedVulkan::vkCmdCopyQueryPoolResults(VkCommandBuffer commandBuffer, VkQ
...
+    if(qpInfo)
+    {
+      VkMemoryBarrier barrier = {
+          VK_STRUCTURE_TYPE_MEMORY_BARRIER,
+          NULL,
+          VK_ACCESS_TRANSFER_WRITE_BIT,
+          VK_ACCESS_TRANSFER_READ_BIT,
+      };
+      ObjDisp(commandBuffer)
+          ->CmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_TRANSFER_BIT,
```
Comment (baldurk) [inline renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp]: These calls now need `Unwrap()`s for the command buffer and buffers, since we're no longer going to do it automatically going through our capture hook.
